include/gabia/secure_socket.hpp

```diff
@@ -0,0 +1,302 @@
+//
+// Copyright (c) 2017 Damian Jarek (damian dot jarek93 at gmail dot com)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+// Official repository: https://github.com/djarek/gabia
+//
+
+#ifndef GABIA_SECURE_SOCKET_HPP
+#define GABIA_SECURE_SOCKET_HPP
+
+#include <gabia/bytes.hpp>
+#include <gabia/crypto/aead.hpp>
+
+#include <boost/asio.hpp>
+#include <boost/beast.hpp>
+#include <boost/endian/arithmetic.hpp>
+
+#define CORO_YIELD BOOST_ASIO_CORO_YIELD
+#define CORO_REENTER BOOST_ASIO_CORO_REENTER
+
+namespace gabia {
+
+namespace asio = boost::asio;
+namespace beast = boost::beast;
+
+template <typename Stream>
+class secure_socket {
+public:
+    using next_layer_type = Stream;
+
+    constexpr static auto max_payload_length = 1024;
+
+    secure_socket(asio::io_service& service) : next_layer_stream{service} {}
+
+    secure_socket(next_layer_type&& next_layer_stream,
+                  crypto::aead_secrets secrets)
+        : next_layer_stream{std::move(next_layer_stream)}, secrets{secrets} {}
```

Comment:
Let me suggest doing away with those constructors and using this signature instead:
```
template<class... Args>
secure_socket(crypto::aead_secrets secrets, Args&&... args)
    : next_layer_stream(std::forward<Args>(args)...)
...
```